### PR TITLE
Scm transmit

### PR DESCRIPTION
**Pull request by dberliner**

### CODE REVIEW

File: drivers/usb/gadget/function/f_scm.c [hunk 1]
```diff
@@ -666,6 +678,34 @@ static void scm_send_int_msg(void *data, int len, struct f_scm *scm_inst)
...
+}
+
+static void scm_send_bulk_msg(struct scm_packet_hdr *hdr, void *data, int len,
+	struct f_scm *scm_inst)
+{
+	struct usb_request *in_req;
+	void *usb_data; 
+	int total_packet_len = sizeof(*hdr) + len;
+
+	printk(KERN_INFO "scm_send_bulk_msg");
```
Comment (dberliner) [inline drivers/usb/gadget/function/f_scm.c]: Remove printks

File: drivers/usb/gadget/function/f_scm.c [hunk 2]
```diff
@@ -709,6 +749,57 @@ static int scm_read_out_cmd(struct f_scm *scm_inst)
...
+	struct f_scm *usb_context)
+{
+	printk(KERN_INFO "scm_process_out_data");
+	/**
+	 *Make sure the packet is big enough for the packet and payload
+	 * (checked in order to avoid reading bad memory)
+	 */
+	if (!packet || len < sizeof(struct scm_packet_hdr) ||
+		len > (sizeof(struct scm_packet_hdr)+packet->hdr.payload_len)) {
+		printk(KERN_INFO "Sanity check failed !%p || %d < %d || %d > %d",
```
Comment (dberliner) [inline drivers/usb/gadget/function/f_scm.c]: Remove printks (global)

File: drivers/usb/gadget/function/f_scm.c [hunk 3]
```diff
@@ -709,6 +749,57 @@ static int scm_read_out_cmd(struct f_scm *scm_inst)
...
+	printk(KERN_INFO "scm_read_out_bulk_cb enter");
+	if (req->buf) {
+		printk(KERN_INFO "scm_read_out_bulk_cb got buf");
+		scm_process_out_data(req->buf, req->actual, req->context);
+		kfree(req->buf);
+	}
+	scm_read_out_bulk(req->context);
+}
+static int scm_read_out_bulk(struct f_scm *scm_inst)
+{
```
Comment (dberliner) [inline drivers/usb/gadget/function/f_scm.c:782]: Rearrange USB functions on top and handling functions below

File: net/scm/xscm.c [hunk 4]
```diff
@@ -209,10 +211,28 @@ static long xaprc00x_wait_for_connect(struct sock *sk, long timeo)
 	return timeo;
 }
 
+/* Realloc a buffer but only copy part of the old one to the begining of the new */
```
Comment (dberliner) [inline net/scm/xscm.c]: Add proper comments and note that shrinking can occur

Comment (dberliner) [reply net/scm/xscm.c]: Make xaprc00x beginning, this isnt a kmalloc function

File: net/scm/xscm.c [hunk 5]
```diff
@@ -225,9 +245,29 @@ void xaprc00x_sock_transmit(int sock_id, void *data, int len)
 	sk = &psk->sk;
 
 	lock_sock(sk);
-	psk->read_cache = krealloc(psk->read_cache,
-		psk->read_cache_len+len, GFP_KERNEL);
-	memcpy(psk->read_cache + psk->read_cache_len, data, len);
+
+	/* Is the cache large enough to fit this data? */
+	free_space = psk->read_cache_alloc_size - psk->read_cache_len;
+	new_len = psk->read_cache_start + psk->read_cache_len + len;
```
Comment (dberliner) [inline net/scm/xscm.c]: Clearer variable names

Comment (dberliner) [reply net/scm/xscm.c]: And a circular buffer with a kern provided structure (as a future issue)

File: drivers/usb/gadget/function/f_scm.c [hunk 6]
```diff
@@ -692,8 +690,6 @@ static void scm_send_bulk_msg(struct scm_packet_hdr *hdr, void *data, int len,
 	void *usb_data; 
 	int total_packet_len = sizeof(*hdr) + len;
 
-	printk(KERN_INFO "scm_send_bulk_msg");
```
Comment (dberliner) [inline drivers/usb/gadget/function/f_scm.c]: Separate printk changes from functiuon

File: net/scm/xscm.c [hunk 7]
```diff
@@ -63,79 +64,76 @@ struct rhashtable g_scm_socket_table;
...
+			1->2 bit 2 snds.
+			2->3 */
+	if ((how & ~SHUTDOWN_MASK) || !how) /* MAXINT->0 */
+		return -EINVAL;
 
-	scm_proxy_close_socket(local_id, g_proxy_context);
+	xaprc00x_sock_side_shutdown_internal(sk, how);
 
+	/* Send shutdown to peer */
+	scm_proxy_close_socket(psk->local_id, g_proxy_context);
```
Comment (dberliner) [inline net/scm/xscm.c:110]: Send CLOSE or SHUTDOWN depending on HOW

Comment (dberliner) [reply net/scm/xscm.c:110]: Make future issue

File: net/scm/xscm.c [hunk 8]
```diff
@@ -63,79 +64,76 @@ struct rhashtable g_scm_socket_table;
...
 }
 
 /**
- * For the proxy to call when the host initiated a close
- * This function will not send anything to the host becuase the sock is already
- * closed and the caller of this function is responsible for sending the ACK.
+ * For the proxy to run when a shutdown is received from the host.
  */
-int xaprc00x_sock_handle_shutdown(int sock_id)
+int xaprc00x_sock_handle_host_side_shutdown(int sock_id, int how)
```
Comment (dberliner) [inline net/scm/xscm.c:117]: Shutdown or close?

File: net/scm/xscm.c [hunk 9]
```diff
@@ -63,79 +64,76 @@ struct rhashtable g_scm_socket_table;
...
 	return 0;
 }
-EXPORT_SYMBOL_GPL(xaprc00x_sock_handle_shutdown);
+EXPORT_SYMBOL_GPL(xaprc00x_sock_handle_host_side_shutdown);
 
-/**
- * Function called when releasing a socket
- */
-static int xaprc00x_sock_release(struct socket *sock)
+static int xaprc00x_sock_side_release(struct socket *sock)
```
Comment (dberliner) [inline net/scm/xscm.c:138]: Verify that this is doing everything that needs to be done

File: net/scm/xscm.c [hunk 10]
```diff
@@ -351,16 +349,24 @@ static int xaprc00x_sock_sendmsg(struct socket *sock,
...
+	sk = sock->sk;
+	psk = (struct xaprc00x_pinfo *)sk;
 
-	if (atomic_read(&psk->state) != SCM_ESTABLISHED) {
-		return -EDESTADDRREQ;
-	}
+	lock_sock(sk);
+
+	if (sk->sk_shutdown & SEND_SHUTDOWN)
+		return -EPIPE;
```
Comment (dberliner) [inline net/scm/xscm.c]: Release sock

File: net/scm/xscm.c [hunk 11]
```diff
@@ -447,23 +453,32 @@ static unsigned int xaprc00x_sock_poll(struct file *file, struct socket *sock,
...
+	sk = sock->sk;
+	psk = (struct xaprc00x_pinfo *)sk;
 	mask = 0;
 
 	sock_poll_wait(file, sk_sleep(sock->sk), wait);
 
+	state = atomic_read(&psk->state);
+
+	/* POLLHUP if and only if both sides are shut down. */
+	if (sk->sk_shutdown == SHUTDOWN_MASK && state == SCM_CLOSE) {
```
Comment (dberliner) [inline net/scm/xscm.c:551]: Wait until all cache data is read

**Review by drbild — APPROVED** [review]
After in-person code reviews, this looks good to me.

I should suggest squashing and merging to the `xaptum_v4.14.47` branch now. Then open a separate branch/PR to rebase the changes into the `xaptum_v5.4.*` branch.

**Pull request merged**